### DISCUSSION TITLE
How to run Staging and Production on Docker Swarm

### DISCUSSION THREAD

**Discussion by CarlosEspejo** [BretFisher/ama#153] [Q&A]
Do you recommend having a staging docker swarm and a production docker swarm? Or is it simple enough to just treat the nodes as compute resources and easily run staging and production side by side on the same Docker swarm?

**Comment by BretFisher** [comment]
One Swarm, like any orchestrator, can run many stacks and services, so technically you could run it all on one Swarm.

However, think of your infrastructure like you think of your code. You test your code before you push it to production, right? So you should do the same with infrastructure. One way to do this is have a test/staging swarm, where you deploy apps for your dev teams to test on. You then can treat this as production-like infrastructure and test docker commands, settings, etc. 

We all make mistakes. Code CI is meant to help with that problem for code, but infrastructure rarely  has its own CI to test changes. I always recommend to ops team to "give your container deployment admin a chance to screw up in something that is not the production solution".
